web/packages/teleport/src/MFAContext/MFAContext.tsx

```diff
@@ -0,0 +1,45 @@
+import { PropsWithChildren, createContext, useCallback, useRef } from 'react';
+import AuthnDialog from 'teleport/components/AuthnDialog';
+import { useMfa } from 'teleport/lib/useMfa';
+import auth, { MfaChallengeScope } from 'teleport/services/auth/auth';
+import { MfaChallengeResponse } from 'teleport/services/mfa';
+
+export interface MfaContextValue {
+  getAdminActionMfaResponse(reusable?: boolean): Promise<MfaChallengeResponse>;
+}
+
+export const MfaContext = createContext<MfaContextValue>(null);
+
+export const MfaContextProvider = ({ children }: PropsWithChildren) => {
+  const allowReuse = useRef(false);
+  const adminMfa = useMfa({
+    req: {
+      scope: MfaChallengeScope.ADMIN_ACTION,
+      allowReuse: allowReuse.current,
+      isMfaRequiredRequest: {
+        admin_action: {},
+      },
+    },
+  });
+
+  const getAdminActionMfaResponse = useCallback(
+    async (reusable: boolean = false) => {
+      allowReuse.current = reusable;
+      return (await adminMfa.getChallengeResponse()) || {}; // return an empty challenge to prevent mfa retry.
+    },
+    [adminMfa, allowReuse]
+  );
+
+  const mfaCtx = {
+    getAdminActionMfaResponse,
+    useMfa,
+  };
+  auth.setMfaContext(mfaCtx);
```

Comment:
This recreates the object and sets it on every re-render

```suggestion
  useEffect(() => {
      const mfaCtx = {
        getAdminActionMfaResponse,
        useMfa,
      };

      auth.setMfaContext(mfaCtx);
  }, [getAdminActionMfaResponse, useMfa]);
```

Reply:
I'm afraid `useEffect` could be called too late. If we must do this (see my other comment) then I would try something like:
```ts
if (!auth.mfaContext) {
  const mfaCtx = {
        getAdminActionMfaResponse,
        useMfa,
      };
  auth.setMfaContext(mfaCtx);
}
```